### ISSUE TITLE
Can I pass my application configuration to Cobra, and modify it once all of my flags have been parsed?

### ISSUE THREAD

**Issue by tomgeorge** [spf13/cobra#2184]
**TLDR: How can I write up my application before starting cobra when my application's configuration depends on the value of a flag?** 

I watched Carolyn Van Slyck's [excellent video](https://youtu.be/eMz0vni6PAw?si=si759LuSqEiVN4uu) at GopherCon 2019 about writing CLIs with Cobra, and read her [example repository](https://github.com/carolynvs/emote) and some of the porter codebase, where the flow of execution is

1. Construct your application via an `Application` struct, which has a `Configuration` that is unmarshalled from viper
2. When constructing your Cobra commands, you make the `Application` a dependency, like so

```go
func buildRootCmd(a *Application) *Application{
  return &cobra.Cmd{}...
}
```

This style of writing Cobra apps is very nice to work with because it becomes much more testable than other implementations that I have tried, where I'm wiring my application together in `PreRun` etc. You also don't need to query the viper store directly.

 **My question is:** how I would go about doing this if I have application dependencies that require a flag value?

Consider a command `foo --github-token abcdefg`. You can't create your top-level `Application` in a fully complete state and pass that to your root command constructor because you need to wait for cobra to parse the flagsets for the various commands. You also cannot modify the `*Application` that is getting passed around because you're not actually passing it by reference, so something like this would fail:

```go
func main() {
	config := cmd.NewConfig()
	cli := github.NewClient(nil).WithAuthToken(config.Data.Github.Token)
	a := cmd.App{
		Config:     config,
		RepoLister: &cmd.GhRepoLister{Cli: cli},
	}
	if err := cmd.NewRootCommand(a).Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

func NewRootCommand(a *App) *cobra.Command {
	rootCmd := &cobra.Command{
		Use:   "cobra-late-bind",
		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
			a.Config.BindFlags(cmd)
                         a.BindServices() // only affects the top-level *Application
			return nil
		},
	}

	rootCmd.PersistentFlags().StringVar(&a.Config.Data.Github.Token, "github-token", "bar", "github token")
	rootCmd.AddCommand(NewFooCommand(a))
	return rootCmd
}

...

func (a *App) BindServices() {
	fmt.Println("Binding services")
	if a.RepoLister == nil {
		a.RepoLister = &GhRepoLister{Cli: github.NewClient(nil).WithAuthToken(a.Config.Data.Github.Token)}
		fmt.Println("Set repolister")
	}
}
```

I think there are two solutions to this:
- Do not allow things like `github.token` to be passed in as flags, and only allow them to be defined in the config file or the environment. Allowing these things to be passed in as flags conflates configuration with options.
- Have the application depend on a set of factories that will create these clients for you (not sure if this will work because I think you would still run into the problem described above where `rootCmds` application pointer updates the factory but `newChildCommand`'s application pointer does not

Has anybody else run into this problem, and solved it in a way that did not feel hacky? 
I have included some example code that reproduces this issue [here](https://github.com/tomgeorge/cobra-late-bind)
